crates/tx5/src/proto.rs

```diff
@@ -0,0 +1,580 @@
+                            return Err(Error::id(
+                                "InvalidMultipartMessageLen",
+                            ));
+                        }
+
+                        if msg_len as usize + self.incoming.len()
+                            > self.want_size
+                        {
+                            return Err(Error::id("ChunkTooLarge"));
+                        }
+
+                        self.incoming.extend_from_slice(&buf[4..]);
+
+                        if self.incoming.len() == self.want_size {
+                            drop(self.grant_permit.take());
+                            self.state = DecodeState::Ready;
+                            Ok(ProtoDecodeResult::Message(std::mem::take(
+                                &mut self.incoming,
+                            )))
+                        } else {
+                            Ok(ProtoDecodeResult::Idle)
+                        }
+                    } else {
+                        Err(Error::id("RecvUnexpectedMultipartMessage"))
+                    }
+                }
+                ProtoHeader::PermitRequest(permit_len) => {
+                    if self.state == DecodeState::Ready {
+                        self.state =
+                            DecodeState::RemoteAwaitingPermit(permit_len);
+                        Ok(ProtoDecodeResult::RemotePermitRequest(permit_len))
+                    } else {
+                        Err(Error::id("RecvUnexpectedPermitRequest"))
+                    }
+                }
+                ProtoHeader::PermitGrant(permit_len) => {
+                    if self.want_remote_grant {
+                        self.want_remote_grant = false;
+                        if let Some(grant_notify) = self.grant_notify.take() {
+                            let _ = grant_notify.send(());
+                        }
+                        Ok(ProtoDecodeResult::RemotePermitGrant(permit_len))
+                    } else {
+                        Err(Error::id("RecvUnexpectedPermitGrant"))
+                    }
+                }
+            }
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn proto_header_encode_decode() {
+        fn check(hdr: ProtoHeader) {
+            let (a, b, c, d) = hdr.encode().unwrap();
+            let res = ProtoHeader::decode(a, b, c, d).unwrap();
+            assert_eq!(hdr, res);
+        }
+
+        for v in 0..32 {
+            check(ProtoHeader::Version(v, [b't', b'x', b'5']));
+        }
+
+        for v in &[0, 42, 0b00011111111111111111111111111111] {
+            check(ProtoHeader::CompleteMessage(*v));
+            check(ProtoHeader::MultipartMessage(*v));
+            check(ProtoHeader::PermitRequest(*v));
+            check(ProtoHeader::PermitGrant(*v));
+        }
+    }
+
+    #[test]
+    fn proto_header_overflow() {
+        assert!(ProtoHeader::Version(u8::MAX, [b't', b'x', b'5'])
```

Comment:
```suggestion
        assert!(ProtoHeader::Version(0b00100000000000000000000000000000, [b't', b'x', b'5'])
```

Reply:
done, but it was a u8: https://github.com/holochain/tx5/pull/77/commits/7608bfa10c4e5abef44ff55bee324c9e389ec590